Node E2E: Add image white list

Random-Liu opened the pull request:
This is part of #29081. Fixes #29155.

As is discussed with @yujuhong in #29155, it is difficult to maintain the prepull image list if it is not enforced. 

This PR added an image white list in the test framework, only images in the white list could be used in the test. If the image is not in the white list, the test will fail with reason:

```
Image "XXX" is not in the white list, consider adding it to CommonImageWhiteList in test/e2e/common/util.go or NodeImageWhiteList in test/e2e_node/image_list.go
```

Notice that if image pull policy is `PullAlways`, the image is not necessary to be in the white list or prepulled, because the test expects the image to be pulled during the test.

Currently, the image white list is only enabled in node e2e, because the image puller in e2e test is not integrated with the image white list yet.

/cc @kubernetes/sig-node

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/32986)

<!-- Reviewable:end -->

Random-Liu committed:
Add image white list, images in white list will be prepulled, and
only images in white list could be used in the test. Currently only
enabled in node e2e test.

yujuhong commented on test/e2e/common/util.go:
```diff
@@ -24,3 +28,20 @@ const (
 
 var CurrentSuite Suite
+
+// CommonImageWhiteList is the list of images used in common test. These images should be prepulled
```
nit: "These images should be prepulled before a tests starts, so that the tests won't fail due image pulling flakes"

Random-Liu replied:
ACK.

yujuhong commented on test/e2e/framework/pods.go:
```diff
@@ -107,6 +113,20 @@ func (c *PodClient) mungeSpec(pod *api.Pod) {
+			// If the image policy is not PullAlways, the image must be in the white list and
+			// pre-pulled.
+			Expect(ImageWhiteList.Has(c.Image)).To(BeTrue(), "Image %q is not in the white list, consider adding it to CommonImageWhiteList in test/e2e/common/util.go or NodeImageWhiteList in test/e2e_node/image_list.go", c.Image)
+			// Do not pull image because the images in white list should have been prepulled.
```
nit: "Do not pull images during the tests because ..."

Random-Liu replied:
ACK

yujuhong commented:
LGTM. Feel free to apply the label after addressing the minor nits.

Random-Liu commented:
@yujuhong Addressed comments.

@dchen1107 has some concern about this PR because image prepull is just a best effort thing, which may not need to be enforced.

However, the image registry is so flaky, we did have some tests failed because of image pulling #28868, #31774 and #31579. And the real image pulling test also failed so much #29259, #28047 that we have to add retry #32990. At least for now prepull is still required, and if the prepull is not enforced, there is no way to maintain the prepull list.

yujuhong commented:
> @dchen1107 has some concern about this PR because image prepull is just a best effort thing, which may not need to be enforced.

Do we have a flag to enable/disable image pre-pulling? We can enforce this policy only if image prepulling is enabled.

Random-Liu commented:
@yujuhong We have a flag.
You are right, if the pre-pull is disabled, we should not enforce it. Will update the PR.

Random-Liu commented:
@yujuhong Changed the framework to only enforce prepull when the `--prepull-images=true`.

yujuhong commented:
LGTM. 

@dchen1107, I think the new change addressed your concern.

Random-Liu commented:
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/32770.

dchen1107 commented:
LGTM
